src/context.zig

```diff
@@ -0,0 +1,457 @@
+const std = @import("std");
+const testing = std.testing;
+const time = std.time;
+const Allocator = std.mem.Allocator;
+
+/// Error types for context operations
+pub const ContextError = error{
+    DeadlineExceeded,
+    Cancelled,
+    InvalidDeadline,
+};
+
+/// Context provides deadline-based cancellation and timeout functionality
+pub const Context = struct {
+    allocator: Allocator,
+    deadline_ns: ?i128, // Nanoseconds since epoch, null means no deadline
+    cancelled: bool,
+    cancel_reason: ?[]const u8,
+    parent: ?*const Context,
+    children: std.ArrayList(*Context),
+    mutex: std.Thread.Mutex,
+
+    /// Create a new root context with optional deadline
+    pub fn init(allocator: Allocator, deadline_ns: ?i128) Context {
+        return Context{
+            .allocator = allocator,
+            .deadline_ns = deadline_ns,
+            .cancelled = false,
+            .cancel_reason = null,
+            .parent = null,
+            .children = std.ArrayList(*Context).init(allocator),
+            .mutex = std.Thread.Mutex{},
+        };
+    }
+
+    /// Create a background context (no deadline, never cancelled)
+    pub fn background(allocator: Allocator) Context {
+        return init(allocator, null);
+    }
+
+    /// Create a context with deadline from duration in milliseconds
+    pub fn withTimeout(allocator: Allocator, timeout_ms: u64) !Context {
+        const now_ns = time.nanoTimestamp();
+        const deadline_ns = now_ns + (@as(i128, timeout_ms) * time.ns_per_ms);
+        return init(allocator, deadline_ns);
+    }
+
+    /// Create a context with absolute deadline
+    pub fn withDeadline(allocator: Allocator, deadline_ns: i128) !Context {
+        const now_ns = time.nanoTimestamp();
+        if (deadline_ns <= now_ns) {
+            return ContextError.InvalidDeadline;
+        }
+        return init(allocator, deadline_ns);
+    }
+
+    /// Create a child context from parent with optional new deadline
+    pub fn withParent(parent: *const Context, deadline_ns: ?i128) !*Context {
+        const child = try parent.allocator.create(Context);
+        child.* = Context{
+            .allocator = parent.allocator,
+            .deadline_ns = if (deadline_ns) |d| blk: {
+                if (parent.deadline_ns) |parent_deadline| {
+                    break :blk @min(d, parent_deadline);
+                } else {
+                    break :blk d;
+                }
+            } else parent.deadline_ns,
+            .cancelled = false,
+            .cancel_reason = null,
+            .parent = parent,
+            .children = std.ArrayList(*Context).init(parent.allocator),
+            .mutex = std.Thread.Mutex{},
+        };
+
+        // Add to parent's children list (this would need proper synchronization in real use)
+        return child;
+    }
+
+    /// Clean up context resources
+    pub fn deinit(self: *Context) void {
+        self.mutex.lock();
```

Comment:
It's not clear to me what this mutex is protecting. It just seems to be the `cancelled` flag. In that case, an atomic might be good enough. Does the mutex here protect anything that's used concurrently? Can one thread call `deinit` while another is calling `cancel()` - that seems problematic even with the mutex.